cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/remote/RemoteLogAppender.java

```diff
@@ -0,0 +1,168 @@
+
+    try {
+      publisher.addMessage(logMessage);
+    } catch (InterruptedException e) {
+      addInfo("Interrupted when adding log message to queue: " + logMessage.getFormattedMessage());
+    }
+  }
+
+  /**
+   * Publisher service to publish logs to log saver.
+   */
+  private final class RemoteLogPublisher extends AbstractLogPublisher<Map.Entry<Integer, byte[]>> {
+
+    private final int numPartitions;
+    private final LoggingEventSerializer loggingEventSerializer;
+    private final LogPartitionType logPartitionType;
+
+    private RemoteLogPublisher(CConfiguration cConf, int queueSize) {
+      super(queueSize, RetryStrategies.fromConfiguration(cConf, "system.log.process."));
+      this.numPartitions = cConf.getInt(Constants.Logging.NUM_PARTITIONS);
+      this.loggingEventSerializer = new LoggingEventSerializer();
+      this.logPartitionType =
+        LogPartitionType.valueOf(cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).toUpperCase());
+    }
+
+    @Override
+    protected Map.Entry<Integer, byte[]> createMessage(LogMessage logMessage) {
+      String partitionKey = logPartitionType.getPartitionKey(logMessage.getLoggingContext());
+      int partition = partition(partitionKey, numPartitions);
+      return new AbstractMap.SimpleEntry<>(partition, loggingEventSerializer.toBytes(logMessage));
+    }
+
+    @Override
+    protected void publish(List<Map.Entry<Integer, byte[]>> logMessages) throws Exception {
+      // Group the log messages by partition and then publish all messages to their respective partitions
+      Map<Integer, List<byte[]>> partitionedMessages = new HashMap<>();
+      for (Map.Entry<Integer, byte[]> logMessage : logMessages) {
+        List<byte[]> messages = partitionedMessages.computeIfAbsent(logMessage.getKey(), k -> new ArrayList<>());
+        messages.add(logMessage.getValue());
+      }
+
+      for (Map.Entry<Integer, List<byte[]>> partition : partitionedMessages.entrySet()) {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        encodeEvents(os, partition.getKey(), partition.getValue());
+        HttpRequest request = remoteClient.requestBuilder(HttpMethod.POST, "process")
+          .addHeader(HttpHeaders.CONTENT_TYPE, "avro/binary")
+          .withBody(ByteBuffer.wrap(os.toByteArray())).build();
+        HttpResponse response = remoteClient.execute(request);
+        // if something went wrong, throw exception to retry
+        if (response.getResponseCode() != HttpURLConnection.HTTP_OK) {
+          throw new IOException(String.format("Could not append logs for partition %s", partition));
+        }
+      }
+    }
+
+    @Override
+    protected void logError(String errorMessage, Exception exception) {
+      // Log using the status manager
+      addError(errorMessage, exception);
+    }
+  }
+
+  private static int partition(Object key, int numPartitions) {
+    return Math.abs(Hashing.md5().hashString(key.toString()).asInt()) % numPartitions;
```

Comment:
is there a reason a md5 is used instead of using key.hashCode()?

Also, how come the key is an Object? Seems like it's only called with a String?

Reply:
yea it should be `String`. This is same as log string partitioner `StringPartitioner`. Added a comment